// @shopify/cli-kit/dist/store.js
import { content, token, debug } from './output.js';
import constants from './constants.js';
import { Abort } from './error.js';
import Conf from 'conf';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const migrations = {};
const schema = {
    appInfo: {
        type: 'array',
        items: {
            type: 'object',
            properties: {
                appId: {
                    type: 'string',
                },
                orgId: {
                    type: 'string',
                },
                storeFqdn: {
                    type: 'string',
                },
            },
        },
    },
};
let _instance;
export function cliKitStore() {
    if (!_instance) {
        throw new Abort("The CLIKitStore instance hasn't been initialized");
    }
    return _instance;
}
export async function initializeCliKitStore() {
    if (!_instance) {
        // eslint-disable-next-line require-atomic-updates
        _instance = new CLIKitStore({
            schema,
            migrations,
            projectName: 'shopify-cli-kit',
            projectVersion: await constants.versions.cliKit(),
        });
    }
}
export class CLIKitStore extends Conf {
    getAppInfo(directory) {
        debug(content `Reading cached app information for directory ${token.path(directory)}...`);
        const apps = this.get('appInfo') ?? [];
        return apps.find((app) => app.directory === directory);
    }
    setAppInfo(options) {
        debug(content `Storing app information for directory ${token.path(options.directory)}:
${token.json(options)}
`);
        const apps = this.get('appInfo') ?? [];
        const index = apps.findIndex((saved) => saved.directory === options.directory);
        if (index === -1) {
            apps.push(options);
        }
        else {
            const app = apps[index];
            apps[index] = {
                appId: options.appId,
                directory: options.directory,
                title: options.title ?? app.title,
                storeFqdn: options.storeFqdn ?? app.storeFqdn,
                orgId: options.orgId ?? app.orgId,
            };
        }
        this.set('appInfo', apps);
    }
    clearAppInfo(directory) {
        debug(content `Clearning app information for directory ${token.path(directory)}...`);
        const apps = this.get('appInfo') ?? [];
        const index = apps.findIndex((saved) => saved.directory === directory);
        if (index !== -1) {
            apps.splice(index, 1);
        }
        this.set('appInfo', apps);
    }
    getTheme() {
        debug(content `Getting theme store...`);
        return this.get('themeStore');
    }
    setTheme(store) {
        debug(content `Setting theme store...`);
        this.set('themeStore', store);
    }
    getSession() {
        debug(content `Getting session store...`);
        return this.get('sessionStore');
    }
    setSession(store) {
        debug(content `Setting session store...`);
        this.set('sessionStore', store);
    }
    removeSession() {
        debug(content `Removing session store...`);
        this.set('sessionStore', '');
    }
}
