// @shopify/cli-kit/dist/api/graphql/functions/app_function_set.js
import { gql } from 'graphql-request';
export const AppFunctionSetMutation = gql `
  mutation AppScriptSet(
    $uuid: String
    $extensionPointName: ExtensionPointName!
    $title: String!
    $description: String
    $force: Boolean
    $schemaMajorVersion: String
    $schemaMinorVersion: String
    $scriptConfigVersion: String
    $configurationUi: Boolean!
    $configurationDefinition: String
    $moduleUploadUrl: String!
    $library: LibraryInput
    $inputQuery: String
    $appBridge: AppBridgeInput
    $apiVersion: String
  ) {
    appScriptSet(
      uuid: $uuid
      extensionPointName: $extensionPointName
      title: $title
      description: $description
      force: $force
      schemaMajorVersion: $schemaMajorVersion
      schemaMinorVersion: $schemaMinorVersion
      scriptConfigVersion: $scriptConfigVersion
      configurationUi: $configurationUi
      configurationDefinition: $configurationDefinition
      moduleUploadUrl: $moduleUploadUrl
      library: $library
      inputQuery: $inputQuery
      appBridge: $appBridge
      apiVersion: $apiVersion
    ) {
      userErrors {
        field
        message
        tag
      }
      appScript {
        uuid
        appKey
        configSchema
        extensionPointName
        title
      }
    }
  }
`;
